implement protos and bhvs

Rim-777 opened the pull request.

am-kantox commented on lib/lister_proto.ex:
```diff
@@ -0,0 +1,41 @@
+
+defimpl Lister, for: Atom do
+  def to_list(atom) do
+    to_string(atom)
```
```suggestion
    atom
    |> to_string()
```

am-kantox commented on lib/lister_proto.ex:
```diff
@@ -0,0 +1,41 @@
+
+defimpl Lister, for: Map do
+  def to_list(map) do
+    map |> Enum.flat_map(fn {k, v} -> [k, v] end)
```
This produces an unexpected (IMHO) result for `%{foo: [:bar, :baz]}`, but ok.

am-kantox commented on lib/lister_proto.ex:
```diff
@@ -0,0 +1,41 @@
+    def to_list(struct) do
+      struct
+      |> Map.from_struct()
+      |> Enum.reduce([], fn pair, list -> [pair | list] end)
+      |> Enum.reverse()
```
```suggestion
      |> Enum.to_list()
```

am-kantox reviewed (approved):
It surely works and accomplishes the task, but it neglects the “logical” difference between protocols and behaviours.

One might implement a behaviour with a protocol (and vice versa,)  but the core difference is there usually exist many implementations of the said behaviour, like all these home-made `GenServer`s.

Like 

```elixir
defmodule Lister do
  @callback as_list(any()) :: list()
end

defmodule LongLister do
  @behaviour Lister
  
  @impl true
  def as_list(%{} = map), do: Enum.to_list(map)
end

defmodule ShortLister do
  @behaviour Lister
  
  @impl true
  def as_list(%{} = map), do: map |> Map.values() |> Enum.join(", ")
end

defmodule Yasha do
  defstruct introducer: LongLister, first: "Yasha", last: "Berhman"

  def introduce_self(%__MODULE__{} = yasha) do
    yasha |> Map.take(~w|first last|a) |> yasha.introducer().as_list()
  end
end
```

And then `Yasha.introduce_self(%Yasha{introducer: ShortLister})` would produce something more concise. Like that.

Rim-777 commented on lib/lister_proto.ex:
```diff
@@ -0,0 +1,41 @@
+
+defimpl Lister, for: BitString do
+  def to_list(string) do
+    String.split(string, "", trim: true)
```
Thank you. very much, Alexey! As I managed to get that Binary always BitSting but not vice versa,  the only thing I cannot get is why, and, most importantly, how it can be used in my actual working life. If so, I will need to give myself an argument as to why I have to learn it deeply.  Could you please shed some light on it?  
Also, could you share how you'd implement the Lister for  BitString?

am-kantox replied:
> why this datatype is required, and, most importantly, how it can be used in my actual working life.

To parse binary protocols, like what was used and is still used in telephony. To parse bits out of bytes. Many protocols have a specification like “first bit is that, then 3 bits are this, etc.” The following picture (taken from the internet) shows it.  

![image](https://github.com/Rim-777/ElixirCourseHomeWork-7/assets/11388982/22e1ffa2-33ac-4053-9360-eb09c251f860)

> Also, could you share how you'd implement the Lister for BitString?

The protocol does not need to be implemented for every type, in the first place. If needed, I’d parse it down to bits.

```elixir
for <<bit::1 <- bitstring>>, do: bit
```

Rim-777 replied:
Thanks!
One thing is still not clear. 
```
defimpl Lister, for: BitString do
  def to_list(bitstring) do
    for <<bit::1 <- bitstring>>, do: bit
  end
end

Lister.to_list("Some string")
[0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1,
 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, ...]
```

from this follows that any Binary will be converted to the list of bits, whereas I expect when I list a normal string e.g. `"some string" ` it to be converted to `["S", "o", "m", "e", " ", "s", "t", "r", "i", "n", "g"]. But this doesn't happen even if I implement the protocol for String,  or Binary too

```
defimpl Lister, for: String do
  def to_list(string) do
    String.split(string, "", trim: true)
  end
end
```

As I understand since the protocol is implemented for BitString, it in turn includes Binary.  
How would you solve it?
`

am-kantox commented:
With the guard `is_binary/1` within BitString implementation, just
implement 2 clauses.

El mar, 2 abr 2024, 22:01, Timo ***@***.***> escribió:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In lib/lister_proto.ex
> <https://github.com/Rim-777/ElixirCourseHomeWork-7/pull/1#discussion_r1548541399>
> :
>
> > @@ -0,0 +1,41 @@
> +defprotocol Lister do
> +  def to_list(term)
> +end
> +
> +defimpl Lister, for: BitString do
> +  def to_list(string) do
> +    String.split(string, "", trim: true)
>
> Thanks!
> One thing is still not clear.
>
> defimpl Lister, for: BitString do
>   def to_list(bitstring) do
>     for <<bit::1 <- bitstring>>, do: bit
>   end
> end
>
> Lister.to_list("Some string")
> [0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1,
>  1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, ...]
>
> from this follows that any Binary will be converted to the list of bits,
> whereas I expect when I list a normal string e.g. "some string" it to be
> converted to `["S", "o", "m", "e", " ", "s", "t", "r", "i", "n", "g"]. But
> this doesn't happen even if I implement the protocol for String, or Binary
> too
>
> defimpl Lister, for: String do
>   def to_list(string) do
>     String.split(string, "", trim: true)
>   end
> end
>
> As I understand since the protocol is implemented for BitString, it in
> turn includes Binary.
> How would you solve it?
> `
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Rim-777/ElixirCourseHomeWork-7/pull/1#discussion_r1548541399>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACW4QNTW62KOEU56WOAJLDDY3MFCTAVCNFSM6AAAAABFTQKKP2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSNZUHEYTSNBVGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Rim-777 commented:
> With the guard `is_binary/1` within BitString implementation, just implement 2 clauses. El mar, 2 abr 2024, 22:01, Timo ***@***.***> escribió:
> […](#)
> ***@***.**** commented on this pull request. ------------------------------ In lib/lister_proto.ex <[#1 (comment)](https://github.com/Rim-777/ElixirCourseHomeWork-7/pull/1#discussion_r1548541399)> : > @@ -0,0 +1,41 @@ +defprotocol Lister do + def to_list(term) +end + +defimpl Lister, for: BitString do + def to_list(string) do + String.split(string, "", trim: true) Thanks! One thing is still not clear. defimpl Lister, for: BitString do def to_list(bitstring) do for <<bit::1 <- bitstring>>, do: bit end end Lister.to_list("Some string") [0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, ...] from this follows that any Binary will be converted to the list of bits, whereas I expect when I list a normal string e.g. "some string" it to be converted to `["S", "o", "m", "e", " ", "s", "t", "r", "i", "n", "g"]. But this doesn't happen even if I implement the protocol for String, or Binary too defimpl Lister, for: String do def to_list(string) do String.split(string, "", trim: true) end end As I understand since the protocol is implemented for BitString, it in turn includes Binary. How would you solve it? ` — Reply to this email directly, view it on GitHub <[#1 (comment)](https://github.com/Rim-777/ElixirCourseHomeWork-7/pull/1#discussion_r1548541399)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACW4QNTW62KOEU56WOAJLDDY3MFCTAVCNFSM6AAAAABFTQKKP2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSNZUHEYTSNBVGA> . You are receiving this because you commented.Message ID: ***@***.***>



> With the guard `is_binary/1` within BitString implementation, just implement 2 clauses.

right, I should've guessed,  thank you!

Rim-777 commented:
> Yasha.introduce_self(%Yasha{introducer: ShortLister}

Огонь! 🔥 Думаю Яша Берхман, кто бы он не был 😂 будет рад!
